extensions/amp-ad/0.1/test/test-amp-ad-3p-impl.js

```diff
@@ -357,4 +359,170 @@ describes.realWin('amp-ad-3p-impl', {
       });
     });
   });
+
+  describe('responsive', () => {
+
+    const VIEWPORT_WIDTH = 300;
+    const VIEWPORT_HEIGHT = 600;
+
+    beforeEach(() => {
+      adConfig['_ping_'].fullWidthHeightRatio = 1.2;
+      win.document.body.removeChild(ad3p.element);
+    });
+
+    describe('#buildCallback', () => {
+      let iframe;
+      let impl;
+      let element;
+
+      function constructImpl(config) {
+        iframe = doc.createElement('iframe');
+
+        config.type = '_ping_';
+        element = createElementWithAttributes(doc, 'amp-ad', config);
+        element.appendChild(iframe);
+        document.body.appendChild(element);
+        impl = new AmpAd3PImpl(element);
+        impl.element.style.display = 'block';
+        impl.element.style.position = 'relative';
+        impl.element.style.top = '101vh';
+
+        // Fix the viewport to a consistent size to that the test doesn't depend
+        // on the actual browser window opened.
+        impl.getViewport().getSize =
+          () => ({width: VIEWPORT_WIDTH, height: VIEWPORT_HEIGHT});
+      }
+
+      it('should do nothing for non-responsive', () => {
+        constructImpl({
+          width: '280',
+          height: '280',
+        });
+        expect(impl.buildCallback()).to.be.undefined;
+      });
+
+      it('should schedule a resize for responsive', () => {
+        constructImpl({
+          width: '100vw',
+          height: '280',
+          'data-auto-format': 'rspv',
+          'data-full-width': '',
+        });
+
+        const callback = impl.buildCallback();
+        expect(callback).to.exist;
+
+         // The returned promise fails for some reason.
+        return callback.then(() => {
+          expect(element.offsetWidth).to.equal(VIEWPORT_WIDTH);
+          expect(element.offsetHeight).to.equal(250);
+        });
+      });
+    });
+    describe('#onLayoutMeasure', () => {
+        // Nested elements to contain the ad. (container contains the ad, and
+        // containerContainer contains that container.)
+      let containerContainer, container;
+      let iframe;
+      let viewer;
+      let element;
+      let impl;
+
+      function buildImpl(config) {
+        containerContainer = doc.createElement('div');
+        container = doc.createElement('div');
+          // Create an element with horizontal margins for the ad to break out
+          // of.
```

Comment:
can we test against body padding as well?

Reply:
Done.